presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java

```diff
@@ -0,0 +1,401 @@
+        readOffset = 0;
+        presentStream = null;
+        dataStream = null;
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(values) + sizeOf(nulls) + sizeOf(outputPositions);
+    }
+
+    private void openRowGroup() throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount) throws IOException
+    {
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        allNulls = false;
+
+        if (outputRequired) {
+            ensureValuesCapacity(positionCount, nullsAllowed && presentStream != null);
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
+        }
+        else {
+            outputPositions = positions;
+        }
+
+        // account memory used by values, nulls and outputPositions
+        systemMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
+        if (dataStream == null && presentStream != null) {
+            streamPosition = readAllNulls(positions, positionCount);
+        }
+        else if (filter == null) {
+            streamPosition = readNoFilter(positions, positionCount);
+        }
+        else {
+            // This branch is inlined because extracting this code into a helper method (readWithFilter)
+            // results in performance regressions on BenchmarkSelectiveStreamReaders.readBooleanNoNull[withFilter]
+            // benchmarks.
+            // The differences are:
+            // - readBooleanNoNull: 0.045 vs.0.029 s/op
+            // - readBooleanNoNullWithFilter: 0.112 vs 0.099 s/op
+            outputPositionCount = 0;
+            byte[] temp = new byte[1];
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    if (nullsAllowed) {
+                        if (outputRequired) {
+                            nulls[outputPositionCount] = true;
+                        }
+                        outputPositions[outputPositionCount] = position;
+                        outputPositionCount++;
+                    }
+                }
+                else {
+                    byte value = dataStream.next();
+                    temp[0] = value;
```

Comment:
this is not very elegant solution; would it make sense to add `testByte` method to `TupleDomainFilter`?